File: C++/Strings/Strings_Day_8_Anagram.cpp
// Anagram
// Difficulty: EasyAccuracy: 44.93%Submissions: 362K+Points: 2
// Given two strings s1 and s2 consisting of lowercase characters. The task is to check whether two given strings are an anagram of each other or not. An anagram of a string is another string that contains the same characters, only the order of characters can be different. For example, act and tac are an anagram of each other. Strings s1 and s2 can only contain lowercase alphabets.

// Note: You can assume both the strings s1 & s2 are non-empty.

// Examples :

// Input: s1 = "geeks", s2 = "kseeg"
// Output: true
// Explanation: Both the string have same characters with same frequency. So, they are anagrams.

#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

bool areAnagrams(string s1, string s2) {
    if (s1.length() != s2.length()) return false; // Early exit if lengths differ

    unordered_map<char, int> char_count;

    // Increment for s1 and decrement for s2
    for (int i = 0; i < s1.length(); i++) {
        char_count[s1[i]]++;
        char_count[s2[i]]--;
    }

    // If any character count is not zero, they're not anagrams
    for (auto it : char_count) {
        if (it.second != 0) {
            return false;
        }
    }

    return true;
}

int main() {
    string s1, s2;
    
    // Input strings from the user
    cout << "Enter the first string: ";
    cin >> s1;
    cout << "Enter the second string: ";
    cin >> s2;

    // Call the function and print the result
    if (areAnagrams(s1, s2)) {
        cout << "The strings are anagrams of each other." << endl;
    } else {
        cout << "The strings are not anagrams of each other." << endl;
    }

    return 0;
}

File: Python/Strings/Strings_Day_8_Anagram.py
# Anagram
# Difficulty: EasyAccuracy: 44.93%Submissions: 362K+Points: 2
# Given two strings s1 and s2 consisting of lowercase characters. The task is to check whether two given strings are an anagram of each other or not. An anagram of a string is another string that contains the same characters, only the order of characters can be different. For example, act and tac are an anagram of each other. Strings s1 and s2 can only contain lowercase alphabets.

# Note: You can assume both the strings s1 & s2 are non-empty.

# Examples :

# Input: s1 = "geeks", s2 = "kseeg"
# Output: true
# Explanation: Both the string have same characters with same frequency. So, they are anagrams.

from collections import Counter

def are_anagrams(s1: str, s2: str) -> bool:
    # Count the frequency of characters in both strings
    s1_map = Counter(s1)
    s2_map = Counter(s2)
    
    # Check if the frequency distributions are the same
    return s1_map == s2_map

# Input strings from the user
s1 = input("Enter the first string: ")
s2 = input("Enter the second string: ")

# Call the function and print the result
if are_anagrams(s1, s2):
    print("The strings are anagrams of each other.")
else:
    print("The strings are not anagrams of each other.")
